training/mochi-1/dataset.py

```diff
@@ -0,0 +1,444 @@
+
+        if (
+            not video_latents_path.exists()
+            or not embeds_path.exists()
+            or not attention_mask_path.exists()
+            or (self.image_to_video and not image_latents_path.exists())
+        ):
+            raise ValueError(
+                f"When setting the load_tensors parameter to `True`, it is expected that the `{self.data_root=}` contains three folders named `video_latents`, `prompt_embeds`, and `prompt_attention_mask`. However, these folders were not found. Please make sure to have prepared your data correctly using `prepare_data.py`. Additionally, if you're training image-to-video, it is expected that an `image_latents` folder is also present."
+            )
+
+        if self.image_to_video:
+            image_latent_filepath = image_latents_path.joinpath(pt_filename)
+        video_latent_filepath = video_latents_path.joinpath(pt_filename)
+        embeds_filepath = embeds_path.joinpath(pt_filename)
+        attention_mask_filepath = attention_mask_path.joinpath(pt_filename)
+
+        if not video_latent_filepath.is_file() or not embeds_filepath.is_file() or not attention_mask_filepath.is_file():
+            if self.image_to_video:
+                image_latent_filepath = image_latent_filepath.as_posix()
+            video_latent_filepath = video_latent_filepath.as_posix()
+            embeds_filepath = embeds_filepath.as_posix()
+            attention_mask_filepath = attention_mask_filepath.as_posix()
+            raise ValueError(
+                f"The file {video_latent_filepath=} or {embeds_filepath=} or {attention_mask_filepath=} could not be found. Please ensure that you've correctly executed `prepare_dataset.py`."
+            )
+
+        images = (
+            torch.load(image_latent_filepath, map_location="cpu", weights_only=True) if self.image_to_video else None
+        )
+        latents = torch.load(video_latent_filepath, map_location="cpu", weights_only=True)
+        embeds = torch.load(embeds_filepath, map_location="cpu", weights_only=True)
+        attention_masks = torch.load(attention_mask_filepath,  map_location="cpu", weights_only=True)
+
+        return images, latents, embeds, attention_masks
+
+
+class VideoDatasetWithResizing(VideoDataset):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+
+    def _preprocess_video(self, path: Path) -> torch.Tensor:
+        if self.load_tensors:
+            return self._load_preprocessed_latents_and_embeds(path)
+        else:
+            video_reader = decord.VideoReader(uri=path.as_posix())
+            video_num_frames = len(video_reader)
+
+            nearest_frame_bucket = min(
+                self.frame_buckets, key=lambda x: abs(x - min(video_num_frames, self.max_num_frames))
+            )
+            if video_num_frames < nearest_frame_bucket:
+                # TODO: we could handle this by padding zero frames or duplicating the existing frames?
```

Comment:
This will actually break our "belief" about the correct number of training steps that the training script will run and messup the epoch calculation since we count on dataset to report correct `__len__`. It is okay for now, but it is more on the user to make sure to specify the correct bucketing frames. If they have videos that do not match the criteria and have enough frames, it should be discarded in dataset preparation phase, so we never reach this case.

In #34, I faced the same issue so what I did was to just make videos that don't match any bucket default to 1 frame (effectively single image). Not sure what the best thing to do is, but I would rather handle these cases during dataset preparation because that makes most sense to me. LMK your thoughts

The change I made in #34 is as follows:

![image](https://github.com/user-attachments/assets/0eac6716-4721-43e8-9ea2-ef412152d0a0)

Reply:
Ah that makes sense. I understand this concern because it resulted in an unexpected error during my trial runs too. I will use your changes in this branch. Thanks!

Reply:
Hmm, I think with this change, additionally, one needs to also specify the frame buckets to be `[1, 84]`? Because, otherwise, we will have `self.buckets={(84, 480, 848): []}` which will ultimately lead to a `KeyNotFoundError` later during bucket sampling?